File: services/price-service/internal/workers/worker.go
package workers

import (
	"context"
	"fmt"
	"os"
	"strings"
	"sync"
	"time"

	"github.com/kosarica/price-service/internal/database/sqlcgen"
	"github.com/kosarica/price-service/internal/jsonb"
	"github.com/kosarica/price-service/internal/taskqueue"
	"github.com/rs/zerolog"
)

var log = zerolog.New(os.Stdout).With().Timestamp().Str("component", "worker").Logger()

// isRetryableError checks if an error should be retried.
// Returns false for data constraint violations that won't be fixed by retry.
func isRetryableError(errMsg string) bool {
	// List of non-retryable database errors
	nonRetryablePatterns := []string{
		"duplicate key value violates unique constraint",
		"violates unique constraint",
		"violates check constraint",
		"violates not-null constraint",
		"violates foreign key constraint",
		"violates exclusion constraint",
	}

	errLower := strings.ToLower(errMsg)
	for _, pattern := range nonRetryablePatterns {
		if strings.Contains(errLower, strings.ToLower(pattern)) {
			return false
		}
	}

	return true
}

type WorkerConfig struct {
	WorkerID   string
	TaskTypes  []string
	MaxTasks   int
	NumWorkers int
	PollDelay  time.Duration
}

// ExtendedHandler is a handler that receives the task queue and task ID for parent-child task support
type ExtendedHandler func(ctx context.Context, payload jsonb.TaskQueuePayload, tq *taskqueue.TaskQueue, taskID string) error

type Worker struct {
	queue            *taskqueue.TaskQueue
	config           WorkerConfig
	handlers         map[string]func(context.Context, jsonb.TaskQueuePayload) error
	extendedHandlers map[string]ExtendedHandler
	stopChan         chan struct{}
	running          chan struct{}
	wg               sync.WaitGroup
}

func New(queue *taskqueue.TaskQueue, config WorkerConfig) *Worker {
	return &Worker{
		queue:            queue,
		config:           config,
		handlers:         make(map[string]func(context.Context, jsonb.TaskQueuePayload) error),
		extendedHandlers: make(map[string]ExtendedHandler),
		stopChan:         make(chan struct{}),
		running:          make(chan struct{}),
	}
}

func (w *Worker) RegisterHandler(taskType string, handler func(context.Context, jsonb.TaskQueuePayload) error) {
	w.handlers[taskType] = handler
}

// RegisterExtendedHandler registers a handler that receives the task queue and task ID
// Use this for handlers that need to schedule child tasks
func (w *Worker) RegisterExtendedHandler(taskType string, handler ExtendedHandler) {
	w.extendedHandlers[taskType] = handler
}

func (w *Worker) Start(ctx context.Context) {
	log.Info().
		Str("component", "worker").
		Str("worker_id", w.config.WorkerID).
		Strs("task_types", w.config.TaskTypes).
		Msg("Starting worker")

	for i := 0; i < w.config.NumWorkers; i++ {
		go w.workerLoop(ctx, i)
	}
}

func (w *Worker) Stop() {
	close(w.stopChan)
	log.Info().
		Str("component", "worker").
		Str("worker_id", w.config.WorkerID).
		Msg("Worker stopping, waiting for in-flight tasks")
	w.wg.Wait()
	log.Info().
		Str("component", "worker").
		Str("worker_id", w.config.WorkerID).
		Msg("Worker stopped")
}

func (w *Worker) workerLoop(ctx context.Context, workerNum int) {
	workerID := fmt.Sprintf("%s-%d", w.config.WorkerID, workerNum)
	log.Info().
		Str("component", "worker").
		Str("worker_id", workerID).
		Msg("Starting worker goroutine")

	ticker := time.NewTicker(w.config.PollDelay)
	defer ticker.Stop()

	for {
		select {
		case <-ctx.Done():
			log.Info().
				Str("component", "worker").
				Str("worker_id", workerID).
				Msg("Worker shutting down")
			return

		case <-w.stopChan:
			log.Info().
				Str("component", "worker").
				Str("worker_id", workerID).
				Msg("Worker received stop signal")
			return

		case <-ticker.C:
			w.processTasks(ctx, workerID)
		}
	}
}

func (w *Worker) processTasks(ctx context.Context, workerID string) {
	claimResult := w.queue.ClaimTasks(ctx, taskqueue.ClaimTasksInput{
		WorkerID:  workerID,
		TaskTypes: w.config.TaskTypes,
		MaxTasks:  w.config.MaxTasks,
	})

	if claimResult.Err != nil {
		log.Error().Err(claimResult.Err).Msg("Failed to claim tasks")
		return
	}

	if len(claimResult.Tasks) == 0 {
		return // No tasks to process
	}

	log.Info().
		Str("component", "worker").
		Str("worker_id", workerID).
		Int("task_count", len(claimResult.Tasks)).
		Msg("Worker claimed tasks")

	for _, task := range claimResult.Tasks {
		w.processTask(ctx, workerID, task)
	}
}

func (w *Worker) processTask(ctx context.Context, workerID string, task taskqueue.ClaimedTask) {
	w.wg.Add(1)
	defer w.wg.Done()

	// Extract runId from payload for logging
	runID := ""
	if task.Payload.RunID != nil {
		runID = *task.Payload.RunID
	}

	// Check for extended handler first, then regular handler
	handler, hasHandler := w.handlers[task.TaskType]
	extHandler, hasExtHandler := w.extendedHandlers[task.TaskType]

	if !hasHandler && !hasExtHandler {
		log.Warn().
			Str("task_type", task.TaskType).
			Msg("No handler for task type")
		w.queue.FailTask(ctx, task.ID, "No handler registered", false)
		return
	}

	log.Info().
		Str("component", "worker").
		Str("worker_id", workerID).
		Str("task_id", task.ID).
		Str("task_type", task.TaskType).
		Str("run_id", runID).
		Msg("Worker processing task")

	// Transition to 'processing' status
	pool := w.queue.GetPool()
	queries := sqlcgen.New(pool)
	err := queries.SetTaskProcessingAny(ctx, task.ID)
	if err != nil {
		log.Error().Err(err).Msg("Failed to mark task as processing")
		w.queue.FailTask(ctx, task.ID, fmt.Sprintf("Status update failed: %v", err), false)
		return
	}

	// Execute the appropriate handler
	var handlerErr error
	if hasExtHandler {
		handlerErr = extHandler(ctx, task.Payload, w.queue, task.ID)
	} else {
		handlerErr = handler(ctx, task.Payload)
	}

	if handlerErr != nil {
		shouldRetry := isRetryableError(handlerErr.Error())
		w.queue.FailTask(ctx, task.ID, handlerErr.Error(), shouldRetry)

		logEvent := log.Error().
			Str("task_id", task.ID).
			Str("run_id", runID).
			Err(handlerErr).
			Bool("will_retry", shouldRetry)

		if !shouldRetry {
			logEvent.Msg("Task failed with non-retryable error (constraint violation)")
		} else {
			logEvent.Msg("Task failed and will retry")
		}
		return
	}

	completeErr := w.queue.CompleteTask(ctx, task.ID, task.Payload)
	if completeErr != nil {
		log.Error().Err(completeErr).Msg("Failed to mark task as completed")
		return
	}

	log.Info().
		Str("component", "worker").
		Str("worker_id", workerID).
		Str("task_id", task.ID).
		Str("run_id", runID).
		Msg("Worker completed task")
}

File: src/lib/taskqueue/worker.ts
import {
	type ClaimedTask,
	claimTasks,
	completeTask,
	failTask,
	startProcessing,
	type TaskType,
} from "./index";

export interface WorkerConfig {
	workerId: string;
	taskTypes?: TaskType[];
	maxTasks?: number;
	pollDelay?: number;
}

export class TaskQueueWorker {
	private config: Required<WorkerConfig>;
	private stopSignal: boolean = false;
	private handlers: Map<TaskType, (task: ClaimedTask) => Promise<void>> =
		new Map();
	private runningTasks: Set<Promise<void>> = new Set();
	private readonly reservedClickHouseSlots = 1;

	constructor(config: WorkerConfig) {
		this.config = {
			workerId: config.workerId,
			taskTypes: config.taskTypes ?? [],
			maxTasks: config.maxTasks ?? 10,
			pollDelay: config.pollDelay ?? 5000,
		};
	}

	registerHandler(
		taskType: TaskType,
		handler: (task: ClaimedTask) => Promise<void>,
	) {
		this.handlers.set(taskType, handler);
	}

	async start() {
		console.log(
			`[WORKER] Starting worker: ${this.config.workerId} (types: ${this.config.taskTypes?.join(", ")})`,
		);

		while (!this.stopSignal) {
			try {
				await this.processTasks();
			} catch (error) {
				console.error("[WORKER] Error processing tasks:", error);
			}

			await this.sleep(this.config.pollDelay);
		}

		console.log(
			`[WORKER] Worker ${this.config.workerId} waiting for in-flight tasks...`,
		);

		await Promise.allSettled(this.runningTasks);

		console.log(`[WORKER] Worker ${this.config.workerId} stopped`);
	}

	async stop() {
		console.log(`[WORKER] Stopping worker ${this.config.workerId}`);
		this.stopSignal = true;
	}

	private async processTasks() {
		const availableSlots = this.config.maxTasks - this.runningTasks.size;
		if (availableSlots <= 0) {
			return;
		}

		const configuredTypes =
			this.config.taskTypes.length > 0 ? this.config.taskTypes : undefined;
		const taskTypes = configuredTypes ?? [];
		const supportsClickHouse = taskTypes.includes("clickhouse");

		let claimedTasks: ClaimedTask[] = [];

		// Claim urgent ClickHouse tasks first to keep prices_current fresh as soon
		// as new parquet snapshots are imported.
		if (supportsClickHouse) {
			const clickhouseTasks = await claimTasks(this.config.workerId, availableSlots, [
				"clickhouse",
			]);
			claimedTasks = claimedTasks.concat(clickhouseTasks);
		}

		const remainingSlots = availableSlots - claimedTasks.length;
		if (remainingSlots <= 0) {
			this.startClaimedTasks(claimedTasks);
			return;
		}

		const nonClickhouseTaskTypes = supportsClickHouse
			? taskTypes.filter((type) => type !== "clickhouse")
			: configuredTypes;
		const nonClickhouseClaimLimit =
			supportsClickHouse && nonClickhouseTaskTypes.length > 0
				? Math.max(remainingSlots - this.reservedClickHouseSlots, 0)
				: remainingSlots;

		if (nonClickhouseClaimLimit > 0) {
			const regularTasks = await claimTasks(
				this.config.workerId,
				nonClickhouseClaimLimit,
				nonClickhouseTaskTypes && nonClickhouseTaskTypes.length > 0
					? nonClickhouseTaskTypes
					: undefined,
			);
			claimedTasks = claimedTasks.concat(regularTasks);
		}

		if (claimedTasks.length === 0) {
			return;
		}

		console.log(
			`[WORKER] Worker ${this.config.workerId} claimed ${claimedTasks.length} tasks (running=${this.runningTasks.size})`,
		);
		this.startClaimedTasks(claimedTasks);
	}

	private startClaimedTasks(tasks: ClaimedTask[]) {
		for (const task of tasks) {
			const taskPromise = this.processTask(task);
			this.runningTasks.add(taskPromise);
			taskPromise.finally(() => {
				this.runningTasks.delete(taskPromise);
			});
		}
	}

	private async processTask(task: ClaimedTask) {
		if (this.stopSignal) {
			console.log(`[WORKER] Worker stopping, skipping task ${task.id}`);
			return;
		}

		const handler = this.handlers.get(task.task_type as TaskType);

		if (!handler) {
			console.warn(`[WARN] No handler for task type: ${task.task_type}`);
			await failTask(task.id, "No handler registered", false);
			return;
		}

		console.log(
			`[WORKER] Processing task ${task.id} (type: ${task.task_type})`,
		);

		try {
			await startProcessing(task.id);

			const TASK_TIMEOUT = 30 * 60 * 1000;
			await Promise.race([
				handler(task),
				new Promise<void>((_, reject) =>
					setTimeout(
						() => reject(new Error("Task execution timeout")),
						TASK_TIMEOUT,
					),
				),
			]);
			await completeTask(task.id);
			console.log(`[WORKER] Completed task ${task.id}`);
		} catch (error) {
			console.error(`[WORKER] Task ${task.id} failed:`, error);
			await failTask(
				task.id,
				error instanceof Error ? error.message : String(error),
				true,
			);
		}
	}

	private sleep(ms: number): Promise<void> {
		return new Promise((resolve) => setTimeout(resolve, ms));
	}
}
